### PR TITLE
improve ability to output yul from internal AST representation

### LINKED ISSUE #310: fix mustache file (5 April)
The code generated in `CodeGenYul.scala` seems to be getting mangled when it gets written out via mustache.  For example, on the empty contract, if you print just before you return from `translateContract`, you get 
```
YulObject(EmptyContract,Code(Block(List(ExpressionStatement(FunctionCall(Identifier(mstore),List(Literal(number,64,int),
Literal(number,128,int)))),
ExpressionStatement(FunctionCall(Identifier(return),List(Literal(number,0,int),
Literal(number,0,int))))))),List(YulObject(EmptyContract_deployed,Code(Block(List(ExpressionStatement(FunctionCall(Identifier(mstore),List(Literal(number,64,int),
Literal(number,128,int)))),
ExpressionStatement(FunctionCall(Identifier(return),List(Literal(number,0,int),
Literal(number,0,int))))))),List(),List())),List())
```
which represents the Yul program

```
object "EmptyContract" {
    code {
            mstore(64,128)
            return(0,0)
    }
    object "EmptyContract_deployed" {
        code {
            mstore(64,128)
            return(0,0)
        }
    }
}
```

but the file that gets written to disk contains

```
object "EmptyContract" {
    code {
            mstore(64,128)
            return(0,0)
    }
    object "EmptyContract_deployed" {
        code {
            return(0,0)
        }
    }
}
```

(which is missing the `mstore` in the `_deployed` version).

### PR DESCRIPTION
**Pull request by ivoysey**
this will eventually resolve #310 but it's very much still in progress. i'm making a PR that i will leave as a draft so that it's easy to see what i've done and what's still unimplemented. travis will fail on this, which is currently correct because i've added a bunch of asserts to fail earlier rather than silently emitting no code.

### CODE REVIEW

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 1]
```diff
@@ -3,21 +3,22 @@ package edu.cmu.cs.obsidian.codegen
...
 
 object CodeGenYul extends CodeGenerator {
 
     // TODO improve this temporary symbol table
-    var tempSymbolTable: Map[String, Int] = Map() // map from field identifiers to index in storage
-    var tempTableIdx = 0 // counter indicating the next available slot in the table
-    var stateIdx = -1    // whether or not there is a state
-    var stateEnumMapping: Map[String, Int] = Map() // map from state name to an enum value
-    var stateEnumCounter = 0  // counter indicating the next value to assign since we don't knon the total num of states
+    var tempSymbolTable : Map[String, Int] = Map() // map from field identifiers to index in storage
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:17]: Scala style guide says no space before colon: https://docs.scala-lang.org/style/types.html

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 2]
```diff
@@ -116,14 +117,11 @@ object CodeGenYul extends CodeGenerator {
...
             Identifier("return"),
-            Seq(Literal(LiteralKind.number,"0","int"),Literal(LiteralKind.number,"0","int"))
-        )
-        statement_seq_deploy = statement_seq_deploy :+ ExpressionStatement(retExpr)
-        statement_seq_runtime = statement_seq_runtime :+ ExpressionStatement(retExpr)
+            Seq(Literal(LiteralKind.number,"0","int"),Literal(LiteralKind.number,"0","int")))
 
         // create runtime object
         val runtime_name = contract.name + "_deployed"
-        val runtime_obj = YulObject(runtime_name, Code(Block(statement_seq_runtime)), Seq(), Seq())
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:126]: Scala style guide says to leave these spaces in: https://docs.scala-lang.org/style/method-invocation.html

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 3]
```diff
@@ -227,45 +225,37 @@ object CodeGenYul extends CodeGenerator {
...
+                    assert(false,"boolean expression in conditional translates to a sequence of expressions")
                     Seq()
                 }
-                scrutinee_yul.apply(0) match {
+                scrutinee_yul.head match {
                     case ExpressionStatement(sye) =>
                         val pos_yul = pos.flatMap(translateStatement)
                         val neg_yul = neg.flatMap(translateStatement)
-                        Seq(Switch(sye,Seq(Case(true_lit , Block(pos_yul)), Case(false_lit, Block(neg_yul)))))
+                        Seq(edu.cmu.cs.obsidian.codegen.Switch(sye,Seq(Case(true_lit , Block(pos_yul)), Case(false_lit, Block(neg_yul)))))
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:258]: also no spaces before commas (https://docs.scala-lang.org/style/method-invocation.html)

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala [hunk 4]
```diff
@@ -128,20 +132,22 @@ case class YulObject (name: String, code: Code, subObjects: Seq[YulObject], data
         var dispatch = false
         var dispatchArray: Array[Case] = Array[Case]()
         var deployCall: Array[Call] = Array[Call]()
-        var memoryInitRuntime: String = ""
+        var memoryInitRuntime : String = ""
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:135]: no space here

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala [hunk 5]
```diff
@@ -128,20 +132,22 @@ case class YulObject (name: String, code: Code, subObjects: Seq[YulObject], data
...
 
         for (s <- obj.code.block.statements) {
             s match {
-                case f: FunctionDefinition => deployFunctionArray = deployFunctionArray :+ new Func(f.yulFunctionDefString())
+                case f: FunctionDefinition =>
+                    print("f is: " + f.toString() + "\n")
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:140]: Be sure to remove the print before merging.

**Review by mcoblenz — CHANGES_REQUESTED** [review]

**Comment by ivoysey** [comment]
As of the EOD on 7 April, we produce the following Yul for the 4 examples in `GanacheTests`, and ganache accepts all of them (https://travis-ci.com/github/mcoblenz/Obsidian/builds/222454704). (But remember we don't check the output of what Ganache does with what we sent off to it.)

```
object "IfThenElse" {
    code {
            mstore(64, 128)
        return(0, 0)
    }
    object "IfThenElse_deployed" {
        code {
            mstore(64, 128)
                let selector := shr(224, calldataload(0))
                switch selector
                    case 0x70a08231 {
                    }
function ifthenelse(){
        switch true
case true {sstore(0, 1)}
case false {sstore(0, 0)}
        leave
}
            return(0, 0)
        }
    }
}
```
```
object "EmptyContract" {
    code {
            mstore(64, 128)
        return(0, 0)
    }
    object "EmptyContract_deployed" {
        code {
            mstore(64, 128)
            return(0, 0)
        }
    }
}
```
```
object "BoolLiteral" {
    code {
            mstore(64, 128)
        return(0, 0)
    }
    object "BoolLiteral_deployed" {
        code {
            mstore(64, 128)
                let selector := shr(224, calldataload(0))
                switch selector
                    case 0x70a08231 {
                    }
function BoolLiteral(){
        sstore(0, true)
        sstore(0, false)
}
            return(0, 0)
        }
    }
}
```
```
object "IntConst" {
    code {
            mstore(64, 128)
        return(0, 0)
    }
    object "IntConst_deployed" {
        code {
            mstore(64, 128)
                let selector := shr(224, calldataload(0))
                switch selector
                    case 0x70a08231 {
                    }
function intconst(){
        sstore(0, 4)
        leave
}
            return(0, 0)
        }
    }
}
```

**Comment by ivoysey** [comment]
@mcoblenz this should now be ready for you to review!

**Comment by ivoysey** [comment]
i made a couple of more changes this morning, but mostly syntactic ones for code clarity. the functionality is the same.

**Review by mcoblenz — APPROVED** [review]

**Pull request merged**